dozer/cogs/filter.py

```diff
@@ -0,0 +1,266 @@
+
+	@guild_only()
+	@has_permissions(manage_guild=True)
+	@filter.command()
+	async def remove(self, ctx, id):
+		"""Remove a pattern from the filter list."""
+		with db.Session() as session:
+			result = session.query(WordFilter).filter(WordFilter.id == id).one_or_none()
+			if result is None:
+				await ctx.send("Filter ID {} not found!".format(id))
+				return
+			if result.guild_id != ctx.guild.id:
+				await ctx.send("That Filter does not belong to this guild.")
+				return
+			result.enabled = False
+			await ctx.send("Filter `{}` with name `{}` deleted.".format(result.id,result.friendly_name))
+			self.load_filters(ctx.guild.id)
+	remove.example_usage = "{prefix}filter remove 7"
+
+	@guild_only()
+	@has_permissions(manage_guild=True)
+	@filter.command(name="dm")
+	async def dm_config(self, ctx, config: bool):
+		"""Set whether filter words should be DMed when used in bot messages"""
+		with db.Session() as session:
+			result = session.query(WordFilterSetting).filter(WordFilterSetting.guild_id == ctx.guild.id,
+															 WordFilterSetting.setting_type == "dm")\
+				.one_or_none()
+			try:
+				before_setting = result.value
+				result.value = config
+			except AttributeError:
+				before_setting = None
+				result = WordFilterSetting(guild_id=ctx.guild.id, setting_type="dm", value=config)
+				session.add(result)
+			await ctx.send("The DM setting for this guild has been changed from {} to {}.".format(before_setting, result.value))
+	dm_config.example_usage = "{prefix}filter dm_config True"
+
+	@guild_only()
+	@filter.group(invoke_without_command=True)
+	async def whitelist(self, ctx):
+		"""List all whitelisted roles for this server"""
+		with db.Session() as session:
+			results = session.query(WordFilterRoleWhitelist).filter(WordFilterRoleWhitelist.guild_id == ctx.guild.id).all()
+			roles_text = ""
+			for db_role in results:
+				role = discord.utils.get(ctx.guild.roles, id=db_role.role_id)
+				roles_text += "{}\n".format(role.name)
+			embed = discord.Embed()
+			embed.title = "Whitlisted roles for {}".format(ctx.guild.name)
+			embed.description = "Anybody with any of the roles below will not have their messages filtered."
+			embed.add_field(name="Roles",value=roles_text or "No roles")
+			await ctx.send(embed=embed)
+	whitelist.example_usage = "{prefix}filter whitelist"
+
+	@guild_only()
+	@has_permissions(manage_roles=True)
+	@whitelist.command(name="add")
+	async def whitelist_add(self, ctx, *, role: discord.Role):
+		"""Add a role to the whitelist"""
+		with db.Session() as session:
+			result = session.query(WordFilterRoleWhitelist).filter(WordFilterRoleWhitelist.role_id == role.id).one_or_none()
+			if result is not None:
+				await ctx.send("That role is already whitelisted.")
+			whitelist_entry = WordFilterRoleWhitelist(guild_id=ctx.guild.id, role_id=role.id)
+			session.add(whitelist_entry)
+			await ctx.send("Whitelisted `{}` for this guild.".format(role.name))
```

Comment:
This should be outside the session's `with`-block, so that errors in sending the message (i.e. permissions errors) don't roll back an otherwise successful transaction.